pycqed/analysis_v2/timedomain_analysis.py

```diff
@@ -1106,6 +1125,283 @@ def get_latex_prob_label(prob_label):
         else:
             return r'$|{}\rangle$'.format(prob_label)
 
+    def _get_single_shots_per_qb(self, raw=False):
+        """
+        Gets single shots from the proc_data_dict and arranges
+        them as arrays per qubit
+        Args:
+            raw (bool): whether or not to return  raw shots (before
+            data filtering)
+
+        Returns: shots_per_qb: dict where keys are qb_names and
+            values are arrays of shape (n_shots, n_value_names) for
+            1D measurements and (n_shots*n_soft_sp, n_value_names) for
+            2D measurements
+
+        """
+        # prepare data in convenient format, i.e. arrays per qubit
+        shots_per_qb = dict()        # store shots per qb and per state
+        pdd = self.proc_data_dict    # for convenience of notation
+        key = 'meas_results_per_qb'
+        if raw:
+            key += "_raw"
+        for qbn in self.qb_names:
+                # if "1D measurement" , shape is (n_shots, n_vn) i.e. one
+                # column for each value_name (often equal to n_ro_ch)
+                shots_per_qb[qbn] = \
+                    np.asarray(list(
+                        pdd[key][qbn].values())).T
+                # if "2D measurement" reshape from (n_soft_sp, n_shots, n_vn)
+                #  to ( n_shots * n_soft_sp, n_ro_ch)
+                if np.ndim(shots_per_qb[qbn]) == 3:
+                    assert self.get_param_value("TwoD", False) == True, \
+                        "'TwoD' is False but single shot data seems to be 2D"
+                    n_vn = shots_per_qb[qbn].shape[-1]
+                    n_vn = shots_per_qb[qbn].shape[-1]
+                    # put softsweep as inner most loop for easier processing
+                    shots_per_qb[qbn] = np.swapaxes(shots_per_qb[qbn], 0, 1)
+                    # reshape to 2D array
+                    shots_per_qb[qbn] = shots_per_qb[qbn].reshape((-1, n_vn))
+                # make 2D array in case only one channel (1D array)
+                elif np.ndim(shots_per_qb[qbn]) == 1:
+                    shots_per_qb[qbn] = np.expand_dims(shots_per_qb[qbn],
+                                                       axis=-1)
+
+        return shots_per_qb
+
+    def _get_preselection_masks(self, presel_shots_per_qb, preselection_qbs=None,
+                                predict_proba=True,
+                                classifier_params=None,
+                                preselection_state_int=0):
```

Comment:
good question: for now  this function "replaces" the use of the data_filter. There are several short comings to the data_filter: does not store the mask of which data is kept or not, and also does not work on individual qubits but is the same for all qubits. I do agree that in the future, it would be better to improve the data_filter capabilities and perhaps merge these together, but this would require significant effort. I added a note in the doc string though.